### PR TITLE
Dynamically added images

### PR DESCRIPTION
**Pull request by alexdrimbe**

### CODE REVIEW

**Comment by XhmikosR** [comment]
The last patch isn't needed.

**Comment by XhmikosR** [comment]
On a side note, @feimosi, please disable Travis when .travis.yml isn't present.

File: src/baguetteBox.js [hunk 1]
```diff
@@ -171,6 +171,10 @@
                 return regex.test(element.href);
             });
             var gallery = [];
+            
+            if(tagsNodeList.length === 0){
```
Comment (XhmikosR) [inline src/baguetteBox.js]: Spaces after if and before the bracket.

File: src/baguetteBox.js [hunk 2]
```diff
@@ -171,6 +171,10 @@
...
+                return
```
Comment (XhmikosR) [inline src/baguetteBox.js]: Missing semicolon.

File: src/baguetteBox.js [hunk 3]
```diff
@@ -156,9 +156,9 @@
 
     function bindImageClickListeners(selector, userOptions) {
         // For each gallery bind a click event to every image inside it
-        var gallery = document.querySelectorAll(selector);
-        galleries.push(gallery);
-        [].forEach.call(gallery, function(galleryElement) {
+        var nodeList = document.querySelectorAll(selector);
```
Comment (XhmikosR) [inline src/baguetteBox.js]: Doesn't help at all.

File: src/baguetteBox.js [hunk 4]
```diff
@@ -178,39 +173,33 @@
...
-        galleries.forEach(function(gallery) {
-            [].forEach.call(gallery, function() {
-                var galleryID = imagesMap.length - 1;
-                [].forEach.call(imagesMap[galleryID], function(imageElement) {
-                    unbind(imageElement, 'click', imagedEventHandlers[galleryID + '_' + imageElement]);
-                });
-                imagesMap.pop();
-            });
-        });
+        // galleries.forEach(function(gallery) {
```
Comment (XhmikosR) [inline src/baguetteBox.js]: Doesn't server any purpose keeping this commenting out; that's why git exists.

Comment (alexdrimbe) [inline src/baguetteBox.js]: It has been refactored and removed on a later commit: 91392cc529264cd604f66b8432b8d0c02d7675e1.

Comment (XhmikosR) [inline src/baguetteBox.js]: That is why git rebase exists. You should rebase and cleanup/squash any related commits.

File: src/baguetteBox.js (hunk 3 above)
Comment (alexdrimbe) [inline src/baguetteBox.js]: I wanted to keep in gallery var the DOM element and the eventHandler. Take a look in f67a3a08c9829196098f8b8ed6a656dbdaa78b84.

Comment (XhmikosR) [inline src/baguetteBox.js]: `gallery` describes exactly what the variable is for. `nodeList` is very generic.

Comment (alexdrimbe) [inline src/baguetteBox.js]: what do you say about galleryNodeList?

Comment (XhmikosR) [inline src/baguetteBox.js]: I guess that's better than plain `nodeList`.

Please remove your last patch from this branch.

File: src/baguetteBox.js (hunk 1 above)
Comment (alexdrimbe) [inline src/baguetteBox.js]: done

File: src/baguetteBox.js (hunk 2 above)
Comment (alexdrimbe) [inline src/baguetteBox.js]: fixed

File: src/baguetteBox.js (hunk 3 above)
Comment (alexdrimbe) [inline src/baguetteBox.js]: done

File: src/baguetteBox.js (hunk 4 above)
Comment (alexdrimbe) [inline src/baguetteBox.js]: I have removed the commented lines from commit.

File: src/baguetteBox.js [hunk 5]
```diff
@@ -151,51 +149,76 @@
...
 
     function unbindImageClickListeners() {
-        galleries.forEach(function(gallery) {
-            [].forEach.call(gallery, function() {
-                var galleryID = imagesMap.length - 1;
-                [].forEach.call(imagesMap[galleryID], function(imageElement) {
-                    unbind(imageElement, 'click', imagedEventHandlers[galleryID + '_' + imageElement]);
-                });
-                imagesMap.pop();
+        for(var selector in data) {
```
Comment (XhmikosR) [inline src/baguetteBox.js]: Please use a space after keywords like `if`, `for`, etc.

Comment (alexdrimbe) [inline src/baguetteBox.js]: done

**Comment by XhmikosR** [comment]
@alexdrimbe: make sure there are not new JSHint issues in your branch.

**Comment by alexdrimbe** [comment]
There are no JSHint issues introduced by the new code.

**Comment by XhmikosR** [comment]
@alexdrimbe: please fetch and rebase,

File: src/baguetteBox.js [hunk 6]
```diff
@@ -148,50 +146,75 @@
         supports.svg = testSVGSupport();
 
         buildOverlay();
+        destroyGalleriesForSelector(selector);
```
Comment (feimosi) [inline src/baguetteBox.js]: Why do we destroy on run() ? I'm not sure I understand the function's responsibility.

File: src/baguetteBox.js [hunk 7]
```diff
@@ -148,50 +146,75 @@
...
         // For each gallery bind a click event to every image inside it
-        var gallery = document.querySelectorAll(selector);
-        galleries.push(gallery);
-        [].forEach.call(gallery, function(galleryElement) {
+        var galleryNodeList = document.querySelectorAll(selector),
+            selectorData = {
+                galleries: [],
+                nodeList: galleryNodeList
+            };
+        data[selector] = selectorData;
```
Comment (feimosi) [inline src/baguetteBox.js]: Could you write it in one-var-per-variable style ?

File: src/baguetteBox.js [hunk 8]
```diff
@@ -148,50 +146,75 @@
...
-        galleries.forEach(function(gallery) {
-            [].forEach.call(gallery, function() {
-                var galleryID = imagesMap.length - 1;
-                [].forEach.call(imagesMap[galleryID], function(imageElement) {
-                    unbind(imageElement, 'click', imagedEventHandlers[galleryID + '_' + imageElement]);
-                });
-                imagesMap.pop();
+        for (var selector in data) {
+            if (data.hasOwnProperty(selector)) {
+                destroyGalleriesForSelector(selector);
```
Comment (feimosi) [inline src/baguetteBox.js]: That's a bit confusing part, we should reword the functions' names. Currently _unbindImageClickListeners_ loops over every selector to invoke _destroyGalleriesForSelector_ which in turn unbinds the listeners.

**Comment by feimosi** [comment]
@alexdrimbe could you summarize this PR? I've already forgotten the essence of the change, besides significant code improvements (btw. great job!).

**Comment by alexdrimbe** [comment]
@feimosi this PR could be summarize as follows: If you dynamically add or remove DOM elements from the gallery and then you call baguetteBox.run( someSelector ) again for the same selector the gallery works fine. Before, if you called 'run' function multiple times for the same selector, the click event was attached multiple times for the same elements and this is a source for memory leak; Now if you call run function multiple times it works properly. 

Before:
before

With the change:
with the change

File: src/baguetteBox.js (hunk 6 above)
Comment (alexdrimbe) [inline src/baguetteBox.js]: When we call run() function multiple times for the same selector we want to destroy any cached data for selector and unbind its listeners.

File: src/baguetteBox.js (hunk 7 above)
Comment (alexdrimbe) [inline src/baguetteBox.js]: done

File: src/baguetteBox.js (hunk 8 above)
Comment (alexdrimbe) [inline src/baguetteBox.js]: I have renamed destroyGalleriesForSelector() to removeFromCache() - because this function unbinds the listeners and removes the selector from the data.
I have also renamed unbindImageClickListeners() to clearCachedData(). 
Tell me if you have a better idea and I will rename them :)

Comment (feimosi) [inline src/baguetteBox.js]: Great, that sounds much better for me :)

**Comment by feimosi** [comment]
Ok, we're good to go with this PR. Really good job @alexdrimbe, thank you! 
Also thanks @XhmikosR for reviewing the code on the way.
